sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -64,6 +64,16 @@ ProgramManager &ProgramManager::getInstance() {
   return GlobalHandler::instance().getProgramManager();
 }
 
+// This function allows for identifying "service" kernels. A SYCL service kernel
+// is a kernel that has not been defined by the user but is instead generated by
+// the SYCL runtime. Service kernel name types must be declared in the
+// sycl::detail::__sycl_service_kernel__ namespace which is exclusively used for
+// this purpose. As such service kernels can be identified by
+// __sycl_service_kernel__ appearing in the mangled kernel name.
+static bool isServiceKernel(const std::string &KernelName) {
+  return KernelName.find("__sycl_service_kernel__") != std::string::npos;
```

Comment:
Checking the start of the string was the original approach, but it did not consider Windows mangling. I am not very familiar with the Windows mangling scheme, but I worry that even if we could determine which mangling scheme is used, having templated classes would break the assumption about the end of the mangled name.

Reply:
It seems, like my yesterday's comment didn't get to post here:
Is it possible to cache this flag?
I believe, using unordered_map will allow for quick access. Also, we can eliminate use of mutexes as the initialization is only needed at boot-strap and there are only read requests during run-time.

Reply:
> I believe, using unordered_map will allow for quick access.

You're thinking adding an `unordered_set` (I think set would do the trick) where `AddImages` caches all found service kernel names. Then `ProgramManager::getSYCLDeviceImagesWithCompatibleState` would check against that rather than calling `isServiceKernel`?

> Also, we can eliminate use of mutexes as the initialization is only needed at boot-strap and there are only read requests during run-time.

Are you referring to `m_KernelIDsMutex`?

Reply:
> You're thinking adding an `unordered_set` (I think set would do the trick) where `AddImages` caches all found service kernel names. Then `ProgramManager::getSYCLDeviceImagesWithCompatibleState` would check against that rather than calling `isServiceKernel`?

Right. This will do the thing.

> Are you referring to `m_KernelIDsMutex`?

Never, never, never. I meant any mutex required for lazy initialization of cache. With the cache initialized at boot-strap there is no place for lazy initialization and, thus, no need for thread-safety mechanism except for ensuring that only read access takes place during run-time.

Reply:
> Right. This will do the thing.

I have implemented caching of the service kernels found in `ProgramManager::AddImages`. This allows for faster lookup later. However, one concern I have is that the only time later that service kernels are looked up is in an assert, so in release builds we will never look up anything in the new cache.

> Never, never, never. I meant any mutex required for lazy initialization of cache. With the cache initialized at boot-strap there is no place for lazy initialization and, thus, no need for thread-safety mechanism except for ensuring that only read access takes place during run-time.

I think this may apply to `m_KernelIDsMutex` as well. That is outside the scope of this PR, so I'll make another PR in which we can discuss further.